rspotify-model/src/idtypes.rs

```diff
@@ -15,323 +16,363 @@
+        _ => return Err(IdError::InvalidPrefix),
+    };
+    let rest = chars.as_str();
+
+    let (tpe, id) = rest
+        .rfind(sep)
+        .map(|mid| rest.split_at(mid))
+        .ok_or(IdError::InvalidFormat)?;
+
+    // Note that in case the type isn't known at compile time,
+    // any type will be accepted.
+    match tpe.parse::<Type>() {
+        Ok(tpe) => Ok((tpe, &id[1..])),
+        _ => Err(IdError::InvalidType),
     }
 }
 
-pub trait PlayableId: Id {}
-pub trait PlayContextId: Id {}
-
 /// This macro helps consistently define ID types.
 ///
-/// * The `$type` parameter indicates what type the ID is made out of (say,
-///   `Artist`, `Album`...) from the enum `Type`.
-/// * The `$name` parameter is the identifier of the struct for that type.
+/// * The `$type` parameter indicates what variant in `Type` the ID is for (say,
+///   `Artist`, or `Album`).
+/// * The `$name` parameter is the identifier of the struct.
+/// * The `$validity` parameter is the implementation of `is_id_valid`.
 macro_rules! define_idtypes {
-    ($($type:ident => $name:ident),+) => {
+    ($($type:ident => {
+        name: $name:ident,
+        validity: $validity:expr
+    }),+) => {
         $(
             #[doc = concat!(
-                "ID of type [`Type::", stringify!($type), "`], made up of only \
-                alphanumeric characters. Refer to the [module-level \
-                docs][`crate::idtypes`] for more information.")
-            ]
+                "ID of type [`Type::", stringify!($type), "`]. The validity of \
+                its characters is defined by the closure `",
+                stringify!($validity), "`.\n\nRefer to the [module-level \
+                docs][`crate::idtypes`] for more information. "
+            )]
+            #[repr(transparent)]
             #[derive(Clone, Debug, PartialEq, Eq, Serialize, Hash)]
-            pub struct $name(String);
+            pub struct $name<'a>(Cow<'a, str>);
+
+            impl<'a> $name<'a> {
+                /// The type of the ID, as a constant.
+                const TYPE: Type = Type::$type;
+
+                /// Only returns `true` in case the given string is valid
+                /// according to that specific ID (e.g., some may require
+                /// alphanumeric characters only).
+                pub fn id_is_valid(id: &str) -> bool {
+                    const VALID_FN: fn(&str) -> bool = $validity;
+                    VALID_FN(id)
+                }
 
-            impl Id for $name {
-                #[inline]
-                fn id(&self) -> &str {
-                    &self.0
+                /// Initialize the ID without checking its validity.
+                ///
+                /// # Safety
+                ///
+                /// The string passed to this method must be made out of valid
+                /// characters only; otherwise undefined behaviour may occur.
+                pub unsafe fn from_id_unchecked<S>(id: S) -> Self
```

Comment:
I am not sure if it's necessary to use `unsafe` statement?

Reply:
I mean, we aren't technically violating memory safety here, but we are violating the integrity of the ID type system. So I thought `unsafe` was a good enough indicator to be careful about its usage. I guess it's weird because we don't actually have `unsafe` code inside it, so we could use a crate like https://docs.rs/unsafe_fn/latest/unsafe_fn/. I would use this new lint, however: https://doc.rust-lang.org/stable/nightly-rustc/rustc_lint/builtin/static.UNSAFE_OP_IN_UNSAFE_FN.html

What do you think?

Reply:
I think it should be `unsafe` as we are really violating type safety in the sense we violate guarantees of id types here. So it's up to programmer to uphold these guarantees, hence it's `unsafe`.